### PR TITLE
Fix some problems in Finduuid.cmake

### PR DESCRIPTION
**Pull request by daritter**
* There is some weird CMake behavior where CMake is unable to check for
  the existence of directories in some cases
* If UUID_LIBRARY is given and _uuid_head_only is not set the
  UUID_LIBRARIES variable was never set, leading to undefined symbols in
  libdavix.so

### CODE REVIEW

**Comment by daritter** [comment]
@amadio could you check this also works for you?

File: cmake/modules/Finduuid.cmake [hunk 1]
```diff
@@ -34,16 +34,18 @@ if(NOT UUID_INCLUDE_DIR)
...
 
-if(EXISTS UUID_INCLUDE_DIR)
+if(EXISTS "${UUID_INCLUDE_DIR}")
   set(UUID_INCLUDE_DIRS ${UUID_INCLUDE_DIR})
   set(CMAKE_REQUIRED_INCLUDES ${UUID_INCLUDE_DIRS})
   check_cxx_symbol_exists("uuid_generate_random" "uuid/uuid.h" _uuid_header_only)
 endif()
 
-if(NOT _uuid_header_only AND NOT UUID_LIBRARY)
+if(NOT _uuid_header_only)
```
Comment (amadio) [inline cmake/modules/Finduuid.cmake]: Why did you remove `AND NOT UUID_LIBRARY`? This prevents the user from setting it to a particular version, as it will be overwritten.

File: cmake/modules/Finduuid.cmake [hunk 2]
```diff
@@ -34,16 +34,18 @@ if(NOT UUID_INCLUDE_DIR)
...
   check_library_exists("uuid" "uuid_generate_random" "" _have_libuuid)
   if(_have_libuuid)
-    set(UUID_LIBRARY "uuid")
+    if(NOT UUID_LIBRARY)
+      set(UUID_LIBRARY "uuid")
+    endif()
     set(UUID_LIBRARIES ${UUID_LIBRARY})
   endif()
 endif()
```
Comment (amadio) [inline cmake/modules/Finduuid.cmake]: Maybe we need an `else` clause here to set `UUID_LIBRARIES` to `UUID_LIBRARY` if that has been set by the user?

File: cmake/modules/Finduuid.cmake (hunk 1 above)
Comment (daritter) [reply cmake/modules/Finduuid.cmake]: No, I set both. The problem in my opinion is that if the user has set `UUID_LIBRARY` then `UUID_LIBRARIES` was never set and that is the variable used to determine link dependencies

So `-DUUID_LIBRARY=some/lib.so` would not work because then `UUID_LIBRARIES` was not set. 

Now the logic is to only set `UUID_LIBRARIES` if the uuid is not only a header only library. And if the user did not set `UUID_LIBRARY ` its set to `uuid`, otherwise the user setting is propagated directly to `UUID_LIBRARIES`

Comment (amadio) [reply cmake/modules/Finduuid.cmake]: Yes, you are right. There is a bug in this file, but I think the solution is adding an `else` as per my comment below, so that when the user sets `UUID_LIBRARY`, it gets propagated to `UUID_LIBRARIES` directly. Otherwise this file will set it to `uuid` independently of what the user sets.

File: cmake/modules/Finduuid.cmake (hunk 2 above)
Comment (amadio) [reply cmake/modules/Finduuid.cmake]: Actually, looking again at these changes, the only problem may be that if the user sets both `UUID_INCLUDE_DIR` and `UUID_LIBRARY`, the latter may be ignored if it's decided that uuid is header only.

File: cmake/modules/Finduuid.cmake [hunk 3]
```diff
@@ -34,16 +34,18 @@ if(NOT UUID_INCLUDE_DIR)
...
-if(EXISTS UUID_INCLUDE_DIR)
+if(EXISTS "${UUID_INCLUDE_DIR}")
   set(UUID_INCLUDE_DIRS ${UUID_INCLUDE_DIR})
   set(CMAKE_REQUIRED_INCLUDES ${UUID_INCLUDE_DIRS})
   check_cxx_symbol_exists("uuid_generate_random" "uuid/uuid.h" _uuid_header_only)
 endif()
 
-if(NOT _uuid_header_only AND NOT UUID_LIBRARY)
+if(NOT _uuid_header_only)
   check_library_exists("uuid" "uuid_generate_random" "" _have_libuuid)
```
Comment (amadio) [inline cmake/modules/Finduuid.cmake]: This should probably be inside the `if(NOT UUID_LIBRARY)`, so it's not checked if `UUID_LIBRARY` is already set.

**Comment by gbitzes** [comment]
Hi, thanks for the pull requests! I'm a bit unsure what to pull, though. #28 fixes the `EXISTS` bug, but as far I understand, removing `AND NOT UUID_LIBRARY` is not entirely correct.

#29 seems more correct to me, but doesn't address the `EXISTS` issue. @daritter could you create one more pull request, which only solves the `EXISTS` bug? This way, I can pull #29 and the new PR, and close this one.

**Comment by daritter** [comment]
Hi @gbitzes,

sure thing, just let me test #29 and I will make a new pr afterwards.

**Comment by amadio** [comment]
@gbitzes Sorry, I did miss the `EXISTS` part... Don't feel obliged to merge #29. If @daritter fixes both things in the new PR, you can merge just that.

**Comment by daritter** [comment]
Sorry for the long wait. I was rather busy with other things. I removed the changes conflicting with #29 from this pr and it now only fixes the `EXISTS` part

**Comment by gbitzes** [comment]
Merged, thanks a lot
